A divider (<!--break-->) can be used to display a preview for posts

pingskills opened the pull request:
Here's my attempt at solving issue #99. Please let me know of any feedback you have.

Here's a screenshot of the home page where the second post listed has a divider tag.
![image](https://cloud.githubusercontent.com/assets/4403168/15076225/f8b139b2-13eb-11e6-8577-66b6f4438852.png)

And here's a screenshot of the full post after the Read more link is clicked:
![image](https://cloud.githubusercontent.com/assets/4403168/15076250/1bc00a3c-13ec-11e6-8a35-91bb1b5a83fd.png)

jonatanklosko commented:
This is really good, the only one thing I would like to add here is some friendly hint below the post textarea informing the writer that he can use this.

jonatanklosko commented on WcaOnRails/app/views/posts/_post.html.erb:
```diff
@@ -26,6 +26,10 @@
+      <%=md post.body_teaser %>
+    <% else %>
+      <%=md post.body_full %>
+    <% end %>
```
```ruby
<%=md render_teaser ? post.body_teaser : post.body_full %>
```

pingskills replied:
Nice. I'll update this.

jonatanklosko commented on WcaOnRails/spec/models/post_spec.rb:
```diff
@@ -8,4 +8,18 @@
+
+  it "displays body teaser and body full when break not present" do
+    post = FactoryGirl.build :post, title: "My Second Post", slug: "my-second-post", body: "This post does not have a preview."
+    expect(post).to be_valid
```
It's not a validation test so I would remove this (in the test above too).

jfly commented:
This looks great! Two thoughts:

- Can we add some hint text for posts.body to let people know about this new feature? You can see an example of how to easily add hints here: https://github.com/jfly/worldcubeassociation.org/blob/9d24719c6a7d4e3a95174c88174aeea7fa13f16f/WcaOnRails/config/locales/en.yml#L71. This is nice, because we actually have 2 forms: `posts/new.html.erb` and `posts/edit.html.erb` (they should probably get combined at some point).
- How do you feel about being more liberal in what we accept? For example: `<!--     break    -->` is something people familiar with html will probably expect to be able to do.

I'm a little confused how the link is showing up on it's own line. It looks great, but I don't see how you achieved that.

Laura-O commented:
Not sure if I understand it right: is there an automatic break after x lines or is this manual?

Sorry if I'm missing something, I'm on my mobile now.

jonatanklosko commented:
@Laura-O It's manual, so you can choose directly when you want to split a post.

Laura-O commented:
Ok, good. Then I have nothing to complain. :-)

pingskills commented:
@jfly - I actually didn't do anything special with the Read more link. At the moment it's up to the author of the post to place the <!--break--> tag such that a line will be there. So in the screenshot the markup was as follows:

Architecto voluptatem dolorem est. Sint et ratione et impedit. Itaque ad illum sit esse atque dolores dolorem.

<!--break-->
This is some interesting text after the break which will only be displayed on the individual post!

jfly commented:
Ahh, I understand. Sounds good!

pingskills commented:
@jfly - just smoke testing and if I put something like

This is the second post with this title!<!-- break --> What will happen?

The markdown ens up being This is the second psot with this title![Read more...](posts/show-preview-2) which it then shows as an image with an alt :(

So I think I'll actually add a couple of newlines before the Read more so this never happens and it will ensure it's always on a new line no matter where the break tag is put.

jfly commented:
Please surround your markdown in ```... ``` so we can  actually see what you're saying :)

jfly commented:
Gah! I failed. Triple back ticks :)

pingskills commented:
The markdown becomes: 
`This is the second post with this title![Read more](/posts/show-preview)`

jfly commented:
Oh, I see! Adding some space sounds like a good plan, then.

pingskills commented:
Thanks for the feedback everyone. I modified the break tag so it can be <!--break--> or <!-- break -->. In fact I used the following regular expression /<!--\s?break\s?-->/ so anything matching that.

I also added a hint to the body textarea as you can see from the following screenshot:
![image](https://cloud.githubusercontent.com/assets/4403168/15092152/f64be1d6-14a4-11e6-9c7f-59c7aa642275.png)

So now the previews look like:
![image](https://cloud.githubusercontent.com/assets/4403168/15092156/0c745948-14a5-11e6-96e0-920a32ebbd20.png)

And the full post like so:
![image](https://cloud.githubusercontent.com/assets/4403168/15092160/1db4c896-14a5-11e6-9bff-b5f9dc99d762.png)

jfly commented on WcaOnRails/app/models/post.rb:
```diff
@@ -5,6 +5,21 @@ class Post < ActiveRecord::Base
   validates :body, presence: true
   validates :slug, presence: true, uniqueness: true
 
+  BREAK_TAG_RE = /<!--\s?break\s?-->/
```
This only allows for 0 or 1 spaces, right? What about a non greedy match? Something like ` *?`

pingskills replied:
Yeah 0 or 1. If we want any number of spaces I don't even think we need a non greedy here because it's just spaces and we end with the -->

So BREAK_TAG_RE = /<!--\s*break\s*-->/ should do the trick I think?

pingskills replied:
`/<!--\s*break\s*-->/`

pingskills commented:
I updated the regex to `/<!--\s*break\s*-->/` and added an extra test with more spaces. Let me know what you think.

jonatanklosko commented:
As @jfly suggested this should be non greedy, consider this part of a post:
```
Something then <!-- break --> and another <!-- break -> here.
```
Of course there's no reason for two breaks, anyway in this case we would miss a part of a post when splitting, wherease using a non greedy matcher we

pingskills commented:
Hi @jonatanklosko,

The greedy regex I used `/<!--\s*break\s*-->/` won't be greedy for the following:

`Something then <!-- break --> and another <!-- break -> here.`

Because we are looking for the start of the tag followed by zero or more whitespace then the word break then zero or more whitespace followed by the end tag. This gives us 2 matches on each of the break tags but not a single greedy match. If I'd used `<!--.*break.*-->/` where we are looking for any character then it would be greedy and we'd miss the middle part of the post. But I think this works well as is. I tested it here: http://rubular.com/

jonatanklosko commented:
Sure, I've realised that and removed my comment almost immediately =)

pingskills commented:
Thanks @jonatanklosko, I got the update by email but thought I'd done something wrong with pushing the new squashed commit that removed the message :)

jfly commented:
Thanks @pingskills. Sounds like @jonatanklosko and I were both suffering from the same bout of crazy =). This LGTM! I'll merge it up now. Could you email results, board, WRC and CC software with the feature announcement?

pingskills commented:
I've just sent out the email about the new feature.

jfly commented:
Just saw it. Thanks, Jeff!

On Sun, May 8, 2016, 17:43 Jeff Plumb <notifications@github.com> wrote:

> I've just sent out the email about the new feature.
>
> —
> You are receiving this because you modified the open/close state.
>
>
> Reply to this email directly or view it on GitHub
> <https://github.com/cubing/worldcubeassociation.org/pull/580#issuecomment-217756164>
>
